relay-common/src/constants.rs

```diff
@@ -330,3 +330,293 @@ impl fmt::Display for SpanStatus {
+pub enum MetricUnit {
+    /// A time duration, defaulting to `"millisecond"`.
+    Duration(DurationUnit),
+    /// Size of information derived from bytes, defaulting to `"byte"`.
+    Information(InformationUnit),
+    /// Fractions such as percentages, defaulting to `"ratio"`.
+    Fraction(FractionUnit),
+    /// user-defined units without builtin conversion or default.
+    Custom(CustomUnit),
+    /// Untyped value without a unit (`""`).
+    None,
+}
+
+impl MetricUnit {
+    /// Returns `true` if the metric_unit is [`None`].
+    pub fn is_none(&self) -> bool {
+        matches!(self, Self::None)
+    }
+}
+
+impl Default for MetricUnit {
+    fn default() -> Self {
+        MetricUnit::None
+    }
+}
+
+impl fmt::Display for MetricUnit {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            MetricUnit::Duration(u) => u.fmt(f),
+            MetricUnit::Information(u) => u.fmt(f),
+            MetricUnit::Fraction(u) => u.fmt(f),
+            MetricUnit::Custom(u) => u.fmt(f),
+            MetricUnit::None => f.write_str("none"),
+        }
+    }
+}
+
+impl std::str::FromStr for MetricUnit {
+    type Err = ParseMetricUnitError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s {
+            "nanosecond" | "ns" => Self::Duration(DurationUnit::NanoSecond),
+            "microsecond" => Self::Duration(DurationUnit::MicroSecond),
+            "millisecond" | "ms" => Self::Duration(DurationUnit::MilliSecond),
+            "second" | "s" => Self::Duration(DurationUnit::Second),
+            "minute" => Self::Duration(DurationUnit::Minute),
+            "hour" => Self::Duration(DurationUnit::Hour),
+            "day" => Self::Duration(DurationUnit::Day),
+            "week" => Self::Duration(DurationUnit::Week),
+
+            "bit" => Self::Information(InformationUnit::Bit),
+            "byte" => Self::Information(InformationUnit::Byte),
+            "kilobyte" => Self::Information(InformationUnit::KiloByte),
+            "kibibyte" => Self::Information(InformationUnit::KibiByte),
+            "megabyte" => Self::Information(InformationUnit::MegaByte),
+            "mebibyte" => Self::Information(InformationUnit::MebiByte),
+            "gigabyte" => Self::Information(InformationUnit::GigaByte),
+            "gibibyte" => Self::Information(InformationUnit::GibiByte),
+            "terabyte" => Self::Information(InformationUnit::TeraByte),
+            "tebibyte" => Self::Information(InformationUnit::TebiByte),
+            "petabyte" => Self::Information(InformationUnit::PetaByte),
+            "pebibyte" => Self::Information(InformationUnit::PebiByte),
+            "exabyte" => Self::Information(InformationUnit::ExaByte),
+            "exbibyte" => Self::Information(InformationUnit::ExbiByte),
+
+            "ratio" => Self::Fraction(FractionUnit::Ratio),
+            "percent" => Self::Fraction(FractionUnit::Percent),
+
+            "" | "unit" | "none" => Self::None,
```

Comment:
Had a discussion offline about if `"unit"` should pattern matched to `Self::None`. The reasoning that this was put in place was to track the unit type `()` that exists in different languages. Does this still make sense? I would rather remove it.

Also worthwhile syncing to see if we can better match OpenTelemetry in general here, particularly around defining dimensionless units, and opting into custom units (like recording X number of Y - for ex with a counter for network packets, 100 (measure) packets (unit)).

https://opentelemetry.io/docs/reference/specification/metrics/semantic_conventions/#instrument-units

> Units should follow the [Unified Code for Units of Measure](http://unitsofmeasure.org/ucum.html) (need more clarification in [#705](https://github.com/open-telemetry/opentelemetry-specification/issues/705)).

- Instruments for utilization metrics (that measure the fraction out of a total) are dimensionless and SHOULD use the default unit 1 (the unity).
- Instruments that measure an integer count of something SHOULD only use [annotations](https://ucum.org/ucum.html#para-curly) with curly braces to give additional meaning without the leading default unit (1). For example, use {packets}, {errors}, {faults}, etc.

Reply:
Thanks, agreed to remove `"unit"` from here. This has not been used actively, so we can safely remove it. 

Other than that, this PR merely moves the existing unit definition and derives new traits. We can discuss units in a new DACI or GitHub issue. Units are definitely not ideal yet, and we're also not able to support _X per Y_ units (e.g. frames per second). Note that this is different from fractions.